File: leetcode/0179-largest-number/solution.py
from functools import cmp_to_key
from typing import List


class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        strs = list(map(str, nums))

        def compare(a: str, b: str) -> int:
            if a + b > b + a:
                return -1
            if a + b < b + a:
                return 1
            return 0

        strs.sort(key=cmp_to_key(compare))

        if strs[0] == "0":
            return "0"

        return "".join(strs)

File: leetcode/0179-largest-number/solution.cpp
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    string largestNumber(vector<int>& nums) {
        vector<string> strs;
        for (int n : nums) {
            strs.push_back(to_string(n));
        }

        sort(strs.begin(), strs.end(), [](const string& a, const string& b) {
            return a + b > b + a;
        });

        if (strs[0] == "0") {
            return "0";
        }

        string result;
        for (const string& s : strs) {
            result += s;
        }
        return result;
    }
};
